how can we use http as well or https with self signed certificate

In the android we can use  ConnectionBuilder}that permits connecting to http links, and ignores certificates for https connections. 
is there any way to use http instead of https.
I am getting this error if I am trying to use http

E/AndroidRuntime( 4341): java.lang.RuntimeException: An error occurred while executing doInBackground()
E/AndroidRuntime( 4341): 	at android.os.AsyncTask$3.done(AsyncTask.java:353)
E/AndroidRuntime( 4341): 	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
E/AndroidRuntime( 4341): 	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
E/AndroidRuntime( 4341): 	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
E/AndroidRuntime( 4341): 	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
E/AndroidRuntime( 4341): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
E/AndroidRuntime( 4341): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
E/AndroidRuntime( 4341): 	at java.lang.Thread.run(Thread.java:764)
E/AndroidRuntime( 4341): Caused by: java.lang.IllegalArgumentException: only https connections are permitted
 And can we use https with self signed certificate

--- a/android/src/main/java/io/crossingthestreams/flutterappauth/FlutterAppauthPlugin.java
+++ b/android/src/main/java/io/crossingthestreams/flutterappauth/FlutterAppauthPlugin.java
@@ -49,7 +49,7 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
     private final int RC_AUTH = 65031;
     private PendingOperation pendingOperation;
     private String clientSecret;
-    private Boolean requireHttps;
+    private boolean allowInsecureConnections;
 
     private FlutterAppauthPlugin(Registrar registrar) {
         this.registrar = registrar;
@@ -105,10 +105,7 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
         final ArrayList<String> promptValues = (ArrayList<String>) arguments.get("promptValues");
         Map<String, String> serviceConfigurationParameters = (Map<String, String>) arguments.get("serviceConfiguration");
         Map<String, String> additionalParameters = (Map<String, String>) arguments.get("additionalParameters");
-        requireHttps = (Boolean) arguments.get("requireHttps");
-        if (requireHttps == null) {
-            requireHttps = true;
-        }
+        allowInsecureConnections = (boolean) arguments.get("allowInsecureConnections");
         return new AuthorizationTokenRequestParameters(clientId, issuer, discoveryUrl, scopes, redirectUrl, serviceConfigurationParameters, additionalParameters, loginHint, promptValues);
     }
 
@@ -135,10 +132,7 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
         final ArrayList<String> scopes = (ArrayList<String>) arguments.get("scopes");
         Map<String, String> serviceConfigurationParameters = (Map<String, String>) arguments.get("serviceConfiguration");
         Map<String, String> additionalParameters = (Map<String, String>) arguments.get("additionalParameters");
-        requireHttps = (Boolean) arguments.get("requireHttps");
-        if (requireHttps == null) {
-            requireHttps = true;
-        }
+        allowInsecureConnections = (boolean) arguments.get("allowInsecureConnections");
         return new TokenRequestParameters(clientId, issuer, discoveryUrl, scopes, redirectUrl, refreshToken, authorizationCode, codeVerifier, grantType, serviceConfigurationParameters, additionalParameters);
     }
 
@@ -232,7 +226,7 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
         }
 
         AppAuthConfiguration.Builder authConfigBuilder = new AppAuthConfiguration.Builder();
-        if (!requireHttps) {
+        if (allowInsecureConnections) {
             authConfigBuilder.setConnectionBuilder(InsecureConnectionBuilder.INSTANCE);
         }
 
@@ -262,7 +256,7 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
         }
 
         AppAuthConfiguration.Builder authConfigBuilder = new AppAuthConfiguration.Builder();
-        if (!requireHttps) {
+        if (allowInsecureConnections) {
             authConfigBuilder.setConnectionBuilder(InsecureConnectionBuilder.INSTANCE);
         }
 
@@ -327,7 +321,7 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
         if (authException == null) {
             if (exchangeCode) {
                 AppAuthConfiguration.Builder authConfigBuilder = new AppAuthConfiguration.Builder();
-                if (!requireHttps) {
+                if (allowInsecureConnections) {
                     authConfigBuilder.setConnectionBuilder(InsecureConnectionBuilder.INSTANCE);
                 }
 

--- a/android/src/main/java/io/crossingthestreams/flutterappauth/FlutterAppauthPlugin.java
+++ b/android/src/main/java/io/crossingthestreams/flutterappauth/FlutterAppauthPlugin.java
@@ -3,6 +3,7 @@ package io.crossingthestreams.flutterappauth;
 import android.content.Intent;
 import android.net.Uri;
 
+import net.openid.appauth.AppAuthConfiguration;
 import net.openid.appauth.AuthorizationException;
 import net.openid.appauth.AuthorizationRequest;
 import net.openid.appauth.AuthorizationResponse;
@@ -48,6 +49,7 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
     private final int RC_AUTH = 65031;
     private PendingOperation pendingOperation;
     private String clientSecret;
+    private Boolean requireHttps;
 
     private FlutterAppauthPlugin(Registrar registrar) {
         this.registrar = registrar;
@@ -103,6 +105,10 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
         final ArrayList<String> promptValues = (ArrayList<String>) arguments.get("promptValues");
         Map<String, String> serviceConfigurationParameters = (Map<String, String>) arguments.get("serviceConfiguration");
         Map<String, String> additionalParameters = (Map<String, String>) arguments.get("additionalParameters");
+        requireHttps = (Boolean) arguments.get("requireHttps");
+        if (requireHttps == null) {
+            requireHttps = true;
+        }
         return new AuthorizationTokenRequestParameters(clientId, issuer, discoveryUrl, scopes, redirectUrl, serviceConfigurationParameters, additionalParameters, loginHint, promptValues);
     }
 
@@ -129,6 +135,10 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
         final ArrayList<String> scopes = (ArrayList<String>) arguments.get("scopes");
         Map<String, String> serviceConfigurationParameters = (Map<String, String>) arguments.get("serviceConfiguration");
         Map<String, String> additionalParameters = (Map<String, String>) arguments.get("additionalParameters");
+        requireHttps = (Boolean) arguments.get("requireHttps");
+        if (requireHttps == null) {
+            requireHttps = true;
+        }
         return new TokenRequestParameters(clientId, issuer, discoveryUrl, scopes, redirectUrl, refreshToken, authorizationCode, codeVerifier, grantType, serviceConfigurationParameters, additionalParameters);
     }
 
@@ -221,8 +231,14 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
             authRequestBuilder.setAdditionalParameters(additionalParameters);
         }
 
+        AppAuthConfiguration.Builder authConfigBuilder = new AppAuthConfiguration.Builder();
+        if (!requireHttps) {
+            authConfigBuilder.setConnectionBuilder(InsecureConnectionBuilder.INSTANCE);
+        }
+
+        AppAuthConfiguration authConfig = authConfigBuilder.build();
         AuthorizationRequest authRequest = authRequestBuilder.build();
-        AuthorizationService authService = new AuthorizationService(registrar.context());
+        AuthorizationService authService = new AuthorizationService(registrar.context(), authConfig);
         Intent authIntent = authService.getAuthorizationRequestIntent(authRequest);
         registrar.activity().startActivityForResult(authIntent, exchangeCode ? RC_AUTH_EXCHANGE_CODE : RC_AUTH);
     }
@@ -245,8 +261,14 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
             builder.setAdditionalParameters(tokenRequestParameters.additionalParameters);
         }
 
+        AppAuthConfiguration.Builder authConfigBuilder = new AppAuthConfiguration.Builder();
+        if (!requireHttps) {
+            authConfigBuilder.setConnectionBuilder(InsecureConnectionBuilder.INSTANCE);
+        }
+
+        AppAuthConfiguration authConfig = authConfigBuilder.build();
         TokenRequest tokenRequest = builder.build();
-        AuthorizationService authService = new AuthorizationService(registrar.context());
+        AuthorizationService authService = new AuthorizationService(registrar.context(), authConfig);
         AuthorizationService.TokenResponseCallback tokenResponseCallback = new AuthorizationService.TokenResponseCallback() {
             @Override
             public void onTokenRequestCompleted(
@@ -304,7 +326,13 @@ public class FlutterAppauthPlugin implements MethodCallHandler, PluginRegistry.A
     private void processAuthorizationData(final AuthorizationResponse authResponse, AuthorizationException authException, boolean exchangeCode) {
         if (authException == null) {
             if (exchangeCode) {
-                AuthorizationService authService = new AuthorizationService(registrar.context());
+                AppAuthConfiguration.Builder authConfigBuilder = new AppAuthConfiguration.Builder();
+                if (!requireHttps) {
+                    authConfigBuilder.setConnectionBuilder(InsecureConnectionBuilder.INSTANCE);
+                }
+
+                AppAuthConfiguration authConfig = authConfigBuilder.build();
+                AuthorizationService authService = new AuthorizationService(registrar.context(), authConfig);
                 AuthorizationService.TokenResponseCallback tokenResponseCallback = new AuthorizationService.TokenResponseCallback() {
                     @Override
                     public void onTokenRequestCompleted(

--- a/android/src/main/java/io/crossingthestreams/flutterappauth/InsecureConnectionBuilder.java
+++ b/android/src/main/java/io/crossingthestreams/flutterappauth/InsecureConnectionBuilder.java
@@ -12,7 +12,7 @@ import java.net.URL;
 
 public class InsecureConnectionBuilder implements ConnectionBuilder {
 
-    public static InsecureConnectionBuilder INSTANCE = new InsecureConnectionBuilder();
+    public static final InsecureConnectionBuilder INSTANCE = new InsecureConnectionBuilder();
 
     private InsecureConnectionBuilder() { }
 

--- a/android/src/main/java/io/crossingthestreams/flutterappauth/InsecureConnectionBuilder.java
+++ b/android/src/main/java/io/crossingthestreams/flutterappauth/InsecureConnectionBuilder.java
@@ -1,0 +1,24 @@
+package io.crossingthestreams.flutterappauth;
+
+import android.net.Uri;
+
+import androidx.annotation.NonNull;
+
+import net.openid.appauth.connectivity.ConnectionBuilder;
+
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+public class InsecureConnectionBuilder implements ConnectionBuilder {
+
+    public static InsecureConnectionBuilder INSTANCE = new InsecureConnectionBuilder();
+
+    private InsecureConnectionBuilder() { }
+
+    @NonNull
+    @Override
+    public HttpURLConnection openConnection(@NonNull Uri uri) throws IOException {
+        return (HttpURLConnection) new URL(uri.toString()).openConnection();
+    }
+}

--- a/lib/src/authorization_request.dart
+++ b/lib/src/authorization_request.dart
@@ -12,7 +12,8 @@ class AuthorizationRequest extends CommonRequestDetails
       Map<String, String> additionalParameters,
       String issuer,
       String discoveryUrl,
-      List<String> promptValues}) {
+      List<String> promptValues,
+      bool requireHttps}) {
     this.clientId = clientId;
     this.redirectUrl = redirectUrl;
     this.scopes = scopes;
@@ -22,6 +23,7 @@ class AuthorizationRequest extends CommonRequestDetails
     this.discoveryUrl = discoveryUrl;
     this.loginHint = loginHint;
     this.promptValues = promptValues;
+    this.requireHttps = requireHttps;
   }
 
   Map<String, dynamic> toMap() {

--- a/lib/src/authorization_request.dart
+++ b/lib/src/authorization_request.dart
@@ -13,7 +13,7 @@ class AuthorizationRequest extends CommonRequestDetails
       String issuer,
       String discoveryUrl,
       List<String> promptValues,
-      bool requireHttps}) {
+      bool allowInsecureConnections: false}) {
     this.clientId = clientId;
     this.redirectUrl = redirectUrl;
     this.scopes = scopes;
@@ -23,7 +23,7 @@ class AuthorizationRequest extends CommonRequestDetails
     this.discoveryUrl = discoveryUrl;
     this.loginHint = loginHint;
     this.promptValues = promptValues;
-    this.requireHttps = requireHttps;
+    this.allowInsecureConnections = allowInsecureConnections;
   }
 
   Map<String, dynamic> toMap() {

--- a/lib/src/authorization_token_request.dart
+++ b/lib/src/authorization_token_request.dart
@@ -15,7 +15,7 @@ class AuthorizationTokenRequest extends TokenRequest
       String issuer,
       String discoveryUrl,
       List<String> promptValues,
-      bool requireHttps})
+      bool allowInsecureConnections: false})
       : super(clientId, redirectUrl,
             clientSecret: clientSecret,
             discoveryUrl: discoveryUrl,
@@ -26,7 +26,7 @@ class AuthorizationTokenRequest extends TokenRequest
             additionalParameters: additionalParameters) {
     this.loginHint = loginHint;
     this.promptValues = promptValues;
-    this.requireHttps = requireHttps;
+    this.allowInsecureConnections = allowInsecureConnections;
   }
 
   Map<String, dynamic> toMap() {

--- a/lib/src/authorization_token_request.dart
+++ b/lib/src/authorization_token_request.dart
@@ -14,7 +14,8 @@ class AuthorizationTokenRequest extends TokenRequest
       Map<String, String> additionalParameters,
       String issuer,
       String discoveryUrl,
-      List<String> promptValues})
+      List<String> promptValues,
+      bool requireHttps})
       : super(clientId, redirectUrl,
             clientSecret: clientSecret,
             discoveryUrl: discoveryUrl,
@@ -25,6 +26,7 @@ class AuthorizationTokenRequest extends TokenRequest
             additionalParameters: additionalParameters) {
     this.loginHint = loginHint;
     this.promptValues = promptValues;
+    this.requireHttps = requireHttps;
   }
 
   Map<String, dynamic> toMap() {

--- a/lib/src/authorization_token_request.dart
+++ b/lib/src/authorization_token_request.dart
@@ -23,10 +23,10 @@ class AuthorizationTokenRequest extends TokenRequest
             scopes: scopes,
             grantType: GrantType.authorizationCode,
             serviceConfiguration: serviceConfiguration,
-            additionalParameters: additionalParameters) {
+            additionalParameters: additionalParameters,
+            allowInsecureConnections: allowInsecureConnections) {
     this.loginHint = loginHint;
     this.promptValues = promptValues;
-    this.allowInsecureConnections = allowInsecureConnections;
   }
 
   Map<String, dynamic> toMap() {

--- a/lib/src/common_request_details.dart
+++ b/lib/src/common_request_details.dart
@@ -23,6 +23,9 @@ class CommonRequestDetails implements Mappable {
   /// Additional parameters to include in the request
   Map<String, String> additionalParameters;
 
+  /// Whether to require HTTPS endpoints (only supported on Android)
+  bool requireHttps;
+
   Map<String, dynamic> toMap() {
     return <String, dynamic>{
       'clientId': clientId,
@@ -31,7 +34,8 @@ class CommonRequestDetails implements Mappable {
       'redirectUrl': redirectUrl,
       'scopes': scopes,
       'serviceConfiguration': serviceConfiguration?.toMap(),
-      'additionalParameters': additionalParameters
+      'additionalParameters': additionalParameters,
+      'requireHttps': requireHttps
     };
   }
 }

--- a/lib/src/common_request_details.dart
+++ b/lib/src/common_request_details.dart
@@ -23,8 +23,8 @@ class CommonRequestDetails implements Mappable {
   /// Additional parameters to include in the request
   Map<String, String> additionalParameters;
 
-  /// Whether to require HTTPS endpoints (only supported on Android)
-  bool requireHttps;
+  /// Whether to allow non-HTTPS endpoints (only applicable on Android)
+  bool allowInsecureConnections;
 
   Map<String, dynamic> toMap() {
     return <String, dynamic>{
@@ -35,7 +35,7 @@ class CommonRequestDetails implements Mappable {
       'scopes': scopes,
       'serviceConfiguration': serviceConfiguration?.toMap(),
       'additionalParameters': additionalParameters,
-      'requireHttps': requireHttps
+      'allowInsecureConnections': allowInsecureConnections
     };
   }
 }

--- a/lib/src/token_request.dart
+++ b/lib/src/token_request.dart
@@ -29,7 +29,8 @@ class TokenRequest with CommonRequestDetails {
       String issuer,
       String discoveryUrl,
       this.authorizationCode,
-      this.codeVerifier})
+      this.codeVerifier,
+      bool requireHttps})
       : assert(
             (issuer != null ||
                 discoveryUrl != null ||
@@ -43,6 +44,7 @@ class TokenRequest with CommonRequestDetails {
     this.additionalParameters = additionalParameters;
     this.issuer = issuer;
     this.discoveryUrl = discoveryUrl;
+    this.requireHttps = requireHttps;
   }
 
   Map<String, dynamic> toMap() {

--- a/lib/src/token_request.dart
+++ b/lib/src/token_request.dart
@@ -30,7 +30,7 @@ class TokenRequest with CommonRequestDetails {
       String discoveryUrl,
       this.authorizationCode,
       this.codeVerifier,
-      bool requireHttps})
+      bool allowInsecureConnections: false})
       : assert(
             (issuer != null ||
                 discoveryUrl != null ||
@@ -44,7 +44,7 @@ class TokenRequest with CommonRequestDetails {
     this.additionalParameters = additionalParameters;
     this.issuer = issuer;
     this.discoveryUrl = discoveryUrl;
-    this.requireHttps = requireHttps;
+    this.allowInsecureConnections = allowInsecureConnections;
   }
 
   Map<String, dynamic> toMap() {

